packages/core/react/src/WalletProvider.tsx

```diff
@@ -15,14 +22,69 @@ export interface WalletProviderProps {
     localStorageKey?: string;
 }
 
+let _userAgent: string | null;
+function getUserAgent() {
+    if (_userAgent === undefined) {
+        _userAgent = globalThis.navigator?.userAgent ?? null;
+    }
+    return _userAgent;
+}
+
+function getIsMobile(adapters: Adapter[]) {
+    const userAgentString = getUserAgent();
+    return getEnvironment({ adapters, userAgentString }) === Environment.MOBILE_WEB;
+}
+
+function getUriForAppIdentity() {
+    const location = globalThis.location;
+    if (location == null) {
+        return;
+    }
+    return `${location.protocol}//${location.host}`;
+}
+
 export function WalletProvider({
     autoConnect,
     localStorageKey = 'walletName',
     wallets: adapters,
     ...props
 }: WalletProviderProps) {
-    const [walletName, setWalletName] = useLocalStorage<WalletName | null>(localStorageKey, null);
-    const adapter = useMemo(() => adapters.find((a) => a.name === walletName) ?? null, [adapters, walletName]);
+    const mobileWalletAdapter = useMemo(() => {
```

Comment:
If on mobile, this either prepares a fresh mobile wallet adapter or uses the one the developer has supplied in the `wallets` array.